[#202, #1081] Authorship: improve file collapsing mechanism

ang-zeyu opened the pull request:
Fixes #1081
Alleviates #202

```
Improve file collapsing in authorship view

The authorship tab relies on several dom querying methods to toggle
whether a file is collapsed from view.
This hurts the performance of file collapsing, and detracts from the
use of vue as RepoSense’s front-end framework, decreasing code
readability.

Let's refactor this to use a new variable in the file object instead,
allowing vue reactivity to take over this behaviour.
This also fixes a bug with files not staying collapsed after changing
the file filter [#1081].

With this change, we are able to conveniently display both the expand
and collapse all buttons at the same time. Let's do so as well.

Occasionally, a small number of files have a large number of line
contributions. This slows the initial loading time at best, and causes
the application to crash at worst.
Let's use the new variable to collapse files with more than
2000 lines of code by default, alleviating this.

```

damithc commented:
It's a good idea to render/load the code 'above the fold' first rather than load everything upfront. But there is no need to wait for scrolling to load the rest?. The page can continue to load the remaining code while the user is still looking at the top of the page.

ang-zeyu commented:
> It's a good idea to render/load the code 'above the fold' first rather than load everything upfront. But there is no need to wait for scrolling to load the rest?. The page can continue to load the remaining code while the user is still looking at the top of the page.

I just tried this, but the issue is that some authors have very large, singular contributions ( e.g. test file updates with > 5k loc ). When lazy loading these files, the client still freezes in the meantime.

We may render the collapsed view for such files on first load instead, in addition to progressively rendering more files at a regular interval ( 1s perhaps ). A good threshold for showing the collapsed view first may be 2k loc.

Alternatively, we can further lazy load each file by itself.

Which option should we go with?

damithc commented:
> I just tried this, but the issue is that some authors have very large, singular contributions ( e.g. test file updates with > 5k loc ). When lazy loading these files, the client still freezes in the meantime.

Perhaps we can ignore these rare cases? The user will be more accepting of slow page loads in such cases. But yes, we can collapse such big files by default, similar to what GitHub does.

ang-zeyu commented:
> Perhaps we can ignore these rare cases? The user will be more accepting of slow page loads in such cases. But yes, we can collapse such big files by default, similar to what GitHub does.

Noted 👍, I'll go with collapsing big files by default first

ang-zeyu commented:
Updated, here's the new behaviour:

A single new file will be rendered every 200ms, and only when the previous file has been fully rendered.

In addition, large files ( > 2000 lines ) are initially displayed with the collapsed view, and its code is not rendered into the dom.

------------
Implementation details :

Rendering:

- The `selectedFiles` attribute is watched ( which is changed when opening an authorship view or changing the filter ), which upon changing calls the `displayMoreFiles` method.
- The method updates the an array of files to be displayed `displayedFiles`, and only if the method was not already executed before, and has yet to complete ( e.g. some files have yet to load, but the user changed the filter again ).
- The callback then recursively calls itself once `nextTick` ( for ensuring the previous dom update has finished ) is fufilled and the throttle timeout ( 200ms ) has elapsed.

Collapsed files:

- I needed to rewrite the way collapsed files were shown, since currently, a simple css style of `display: none` is applied over collapsed files. This causes the dom for these files to be rendered still.
  - I removed the `document.getElementsByClassName` approach in favour of an additional variable `file.active`, and rewrote the "toggle collapse" functions to trigger an additional cache flag `file.wasLoaded` .
  - Removed some old methods as a result ( `getSelectedFiles` - which did not have any usages before either, and `updateCount` )

------------
Some decisions that need more reviewing:
- The value of the hardcoded throttle duration - is 200ms too big / small?
- Whether we should lazily load large files ( which are likely to be non-functional though ) over collapsing them
- :grey_question:

ang-zeyu commented:
Ready for review.

The new behaviour:

A single new file will be rendered every 300ms, and only when the previous file has been fully rendered.

In addition, large files ( > 2000 lines ) are initially displayed with the collapsed view, and its code is not rendered into the dom until the user expands it.

Clicking `show all file details` or expanding a single file is still synchronous however.

------------

For Collapsed files:

I needed to rewrite the way collapsed files were shown, since currently, only a css style of `display: none` is applied over collapsed files. This causes the dom for these files to be rendered still.
- I removed the `document.getElementsByClassName` approach in favour of an additional variable `file.active`, and rewrote the "toggle collapse" functions to trigger an additional cache flag `file.wasCodeLoaded`.
- I also removed some unnecessary methods as a result ( `getSelectedFiles` - which did not have any usages before either, and `updateCount` )

damithc commented:
Nice work. 
When I check the Netlify preview for this PR, the first loading of the code view takes quite long and there is no `loading...` indicator. The subsequent loadings (for other authors) are quite fast though. I wonder if that's a netlify-specific problem or something to do with the change in this PR.

Also, should we keep the `loading...` indicator until the first file shows up?

ang-zeyu commented:
> When I check the Netlify preview for this PR, the first loading of the code view takes quite long and there is no `loading...` indicator. The subsequent loadings (for other authors) are quite fast though. I wonder if that's a netlify-specific problem or something to do with the change in this PR.

It may be the large amount of animated spinners ( one per file ), or some other things.
I'll try to micro optimize it!

> Also, should we keep the `loading...` indicator until the first file shows up?

👍

ang-zeyu commented:
Updated,
Its about 1s ( 7 down to 6s ) faster to first interaction for a 2000 line author in the netlify preview, and should scale the more lines there are.
The previous timings was simply due to showing 1 animated spinner per file, which was much more expensive than I expected.

0blivious commented on frontend/src/tabs/authorship.pug:
```diff
@@ -70,8 +70,8 @@
           )
             i.button.fas.fa-user-edit
-        pre.hljs
+        pre.hljs(v-if="doLoadFile(file)" v-show="file.active")
```
Why v-show and v-if together? Also, the attribute `wasCodeLoaded` seems not useful.

ang-zeyu replied:
Hi!
`v-if` is to prevent vue from unloading it from the dom ( caching it essentially ) once it has been loaded for the first time.
In here, `doLoadFile` returns `wasCodeLoaded || file.active`, and `wasCodeLoaded` is set to true and remains permenantly so once the file is set to `active` for the first time, and has been loaded 
( either by `displayMoreFiles` or the user re-expanding a by-default collapsed file )

`v-show` is just to toggle `display: none` on/off on it

0blivious replied:
I understand. But isn't this what v-show does? It make the elements remain in the DOM and are only hidden by CSS.

ang-zeyu replied:
We don't want all of the code lines' DOM to be loaded first ( rendered ) though, which is the purpose of this PR, hence we also need `v-if` to control the loading.

0blivious replied:
Just condition on `wasCodeLoaded` then. Will there be any cases that a file is active but not loaded?

ang-zeyu replied:
Yup, when the authorship view is initially loaded, `wasCodeLoaded` is false for all files.
Files with < 2000 loc are also not `active` in addition as discussed here https://github.com/reposense/RepoSense/pull/1065#issuecomment-583850317

`wasCodeLoaded` is toggled bit by bit for each file, which loads the page asynchronously.
It's also switched on when the user expands files with >= 2000 loc ( by the expand all button or toggling the file individually )

ang-zeyu replied:
@0blivious I see your point now, thanks!

Just to clarify again, you meant that only the `||` in `file.wasCodeLoaded || file.active` is unecessary right?
ie. It should just be `file.wasCodeLoaded`

0blivious replied:
Yes, you can do sth like v-if="file.wasCodeLoaded"

0blivious replied:
Yes, sth like v-if="file.wasCodeLoaded"

jamessspanggg commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -32,6 +32,8 @@ window.vAuthorship = {
     return {
       isLoaded: false,
       files: [],
+      isLoadingMoreFiles: false,
```
one question: is `isLoadingMoreFiles` the inverse of `hasAllLoaded`?

ang-zeyu replied:
yup, except `hasAllLoaded` is only toggled once to avoid triggering too many dom updates from `isLoadingMoreFiles`

jamessspanggg commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -171,6 +181,9 @@ window.vAuthorship = {
           out.lineCount = lineCnt;
+          out.active = lineCnt <= COLLAPSED_VIEW_LINE_COUNT_THRESHOLD;
+          out.wasCodeLoaded = false;
+          out.isLoading = true;
```
Perhaps `shouldLoad` / `shouldRender` may be a better variable name instead of `isLoading`?

ang-zeyu commented:
Realised some additional things: 
1. If the user resorts the files before the loading has finished, the loading spinner would still be at the bottom, but files would pop in throughout since the loading order follows the new sorted order. Should we show a small `loading...` text in place of a spinner per file instead? ( as was before, but text should be less expensive ) ( there's also an added benefit of being able to remove one or two state variables )

2. For authors with a small number of total loc, it would be tiresome to wait for all the files to load. ( though the user can click expand all button to do so ) We could restrict progressive loading to files with a minimal amount of lines of code as well ( 50 maybe ), which should give a good balance of TTI and time spent waiting for files to be lazily loaded

what do you guys think? @jamessspanggg @0blivious @damithc

damithc commented:
> what do you guys think? @jamessspanggg @0blivious @damithc

No objections from me.

0blivious commented on frontend/src/tabs/authorship.pug:
```diff
@@ -1,8 +1,8 @@
-    a(v-if="activeFilesCount === 0", v-on:click="expandAll(true)") show all file details
-    a(v-else, v-on:click="expandAll(false)") hide all file details
+    a(v-if="activeFilesCount < this.selectedFiles.length", v-on:click="expandAll()") show all file details
+    a(v-else, v-on:click="collapseAll()") hide all file details
```
Since we are using vue, is it better we just watch the activeFilesCount instead of implementing 2 click functions? Also, why revamp the previous logic for button change? I feel display `hide all` as long as there is at least one active file would be better since the first thing I do when open the authorship view is usually to hide all file details.

ang-zeyu replied:
> Since we are using vue, is it better we just watch the activeFilesCount instead of implementing 2 click functions?

Hmm I don't follow, isn't vue already watching `activeFilesCount` implicitly and showing/hiding the `a(...)`'s as needed with the `v-if`'s?


> Also, why revamp the previous logic for button change? I feel display hide all as long as there is at least one active file would be better since the first thing I do when open the authorship view is usually to hide all file details.

Fixed :+1:. Only meant to change the condition for the first `a(...)`

0blivious commented:
> what do you guys think? @jamessspanggg @0blivious @damithc

I agree, would prefer a single line of text instead of animation.

Also, can you link this PR to #1081 in your commit message?

jamessspanggg commented:
> what do you guys think? @jamessspanggg @0blivious @damithc

Sounds good to me. I prefer the `loading...` text too

ang-zeyu commented:
Updated, removed `isLoading` and `hasLoadedAll` flags.

For #202,
I tried loading the 400k loc commit as well, it works initially, but still crashes once you expand the 200k loc file. 
Will leave it at "alleviate", since we can at least now view the smaller files though.

jamessspanggg commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -99,19 +102,24 @@ window.vAuthorship = {
 
-    updateCount() {
-      this.activeFilesCount = document.getElementsByClassName('file active').length;
+    toggleCollapsed(file) {
```
perhaps `toggleFile` would be a better method name? as the file may not be collapsed

ang-zeyu replied:
Would `toggleFileActive` be fine?

jamessspanggg replied:
how about `toggleFileActiveProperty`?

ang-zeyu replied:
Sounds good :+1:

jamessspanggg commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -319,6 +351,7 @@ window.vAuthorship = {
     this.initiate();
     this.setInfoHash();
   },
+
```
maybe the new line is not necessary

ang-zeyu replied:
Hmm, most of the "sections" are currently seperated by newlines though.
e.g. between `watch` and `methods` and `components`.
Not sure if github is displaying this correctly, I wouldn't mind changing it otherwise though

jamessspanggg replied:
Alright, then it is fine

jamessspanggg commented on frontend/src/static/css/v_authorship.scss:
```diff
@@ -2,6 +2,14 @@
 /* Authorship */
 #tab-authorship {
+  .toolbar {
+    > a {
```
may I ask why this CSS change is needed?

ang-zeyu replied:
for `show all file details` and `hide all file details` to show up on top / below of each other.
I changed the condition for `show all file details` to `activeFilesCount > 0` from `activeFilesCount === 0`, which can be helpful since we now collapse files with > 2000 loc by default.

jamessspanggg replied:
ahhh okay, @damithc are u ok with the following, where if there's at least one file that is collapsed and at least one file that is showed, then the `show all file details` and `hide all file details` options will appear.

<img width="869" alt="Screen Shot 2020-02-20 at 11 23 46 PM" src="https://user-images.githubusercontent.com/32864116/74949018-18d90900-5438-11ea-968e-48456adee981.png">

damithc replied:
Yes I'm OK with it.

ang-zeyu commented:
Updated `toggleCollapsed` naming

0blivious commented:
@ang-zeyu can fix the conflicts?

ang-zeyu commented:
> @ang-zeyu can fix the conflicts?

:+1: Updated

ang-zeyu commented:
Resolved minor conflict with #1074

ang-zeyu commented:
Resolved some more conflicts with merged PRs

ang-zeyu commented:
> there seems to be some memory leak and heavier memory usage compare to master branch

Thanks for testing this!

What was the method used to produce the leak in your example?
I tried switching between authorship tabs and different types of tabs, but wasn't able to produce it exclusively on this PR ( 'master' has it as well ).

I've updated this with a fix for another potential memory leak though ( clearing the timeout ), but the leak still occurs with it.

( PR preview of [1169](https://deploy-preview-1169--reposense.netlify.com/) after switching through some amount of authorship tabs )
![memoryleak](https://user-images.githubusercontent.com/3306138/77200040-c8cb9000-6b24-11ea-8df1-aceb18f5af69.png)

Memory usage wise, I looked at the usage after entering the page and opening an author with ~2000 loc. In both cases the usages was within error of margin ( ~5 mb ); Could the increased observed usage previously be due to the memory leak?

eugenepeh commented:
>What was the method used to produce the leak in your example?

Memory leak observed was from constantly switching between the authorship view of different authors. 

Although there are some leakage from master branch, it wasn't as severe as the PR.
After some times of repetition, the PR's memory usage doubled compare to master branch.

This has happened before in the past where hljs were used inadequately; perhaps you can look into that area. 

Meanwhile, can other reviewers confirm this as well? In case I am the only one who experience degradation.

ang-zeyu commented:
> Memory leak observed was from constantly switching between the authorship view of different authors.
> 
> Although there are some leakage from master branch, it wasn't as severe as the PR.
> After some times of repetition, the PR's memory usage doubled compare to master branch.

Ah I see. It might be due to the potential memory fix I just updated this with. My timings between switching tabs weren't perfectly consistent before as well, which may have led to different results from yours.
I'm actually getting consistently lower heap usage on this netlify preview versus master now otherwise, which makes sense since we aren't rendering everything at once.

@jamessspanggg @0blivious 
would appreciate if you could help double check if the leak in this pr is more serious than master!

0blivious commented:
For my case it is a bit more serious than master, also the JS heap usage tend to be slightly higher (20MB more).

ang-zeyu commented:
> For my case it is a bit more serious than master, also the JS heap usage tend to be slightly higher (20MB more).

Thanks for testing it out!
I can now reproduce it strangely, and its actually + ~30% usage after a while, which seems rather significant.

I have a solution in mind, being to move the timer logic to the main application ( since it stays in memory all the time, it should prevent leaks from this ), but I think this is something better done after projectifying the frontend with better global state management.

In the meantime, I could repurpose this PR to preserve these behaviours
- refactoring of the `document.getElementsByClassName('...')` method of toggling file collapse to vue data variables to decrease reliance on dom querying
- showing both the `show all files` and `hide all files` button appropriately

Would these changes be fine on its own? @jamessspanggg @0blivious

0blivious commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -40,14 +40,13 @@ window.vAuthorship = {
   watch: {
     selectedFiles() {
-      setTimeout(this.updateCount, 0);
+      this.displayMoreFiles();
```
unresolved function

ang-zeyu replied:
removed

ang-zeyu replied:
thanks for re-reviewing this as well!

I suppose the changes mentioned above are fine on their own?
I've updated the commit message, title and linked issues to reflect, let me know otherwise!

0blivious replied:
yes I feel the current implementation are fine

ang-zeyu commented:
@jamessspanggg could I get your review here again? 

The lazy rendering parts are removed and only the following parts are preserved:
- refactoring of the `document.getElementsByClassName('...')` method of toggling file collapse to vue data variables to decrease reliance on dom querying
- showing both the `show all files` and `hide all files` button appropriately
- collapsing files > 2000 loc by default ( alleviating #202 )

As a sidenote my all my PRs seem to be passing on travis currently https://travis-ci.org/github/reposense/RepoSense/pull_requests?utm_medium=notification&utm_source=github_status but the status isn't getting reported back to github o_o

jamessspanggg commented:
> As a sidenote my all my PRs seem to be passing on travis currently https://travis-ci.org/github/reposense/RepoSense/pull_requests?utm_medium=notification&utm_source=github_status but the status isn't getting reported back to github o_o

There are times where it happens to my PRs as well. Not sure what causes it :/

ang-zeyu commented:
Thanks for the re-review!

> There are times where it happens to my PRs as well. Not sure what causes it :/

Ahh ok, will try triggering a rebuild tomorrow again. ( been triggering for two days already tho =/ )

ang-zeyu commented:
> Clean fix. Can you also explore to apply the same fix to all other methods that use `toggleNext` function?

👍, seems like only `zoom.pug` is still using it. Should it be in a separate PR?

fzdy1914 commented:
> > Clean fix. Can you also explore to apply the same fix to all other methods that use `toggleNext` function?
> 
> 👍, seems like only `zoom.pug` is still using it. Should it be in a separate PR?

Yes, please do it in a separate PR.
